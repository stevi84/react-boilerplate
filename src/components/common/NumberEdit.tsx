import { TextField } from '@mui/material';
import { useState } from 'react';
import { useField } from 'formik';
import { formatNumber, parseNumber } from '../../globals/Formatters';
import { Locale } from '../../globals/Translations';
import { useTranslation } from 'react-i18next';

interface NumberEditProps {
  id: string;
  name: string;
  label: string;
}

export const NumberEdit = (props: NumberEditProps) => {
  const { id, name, label } = props;

  const [field, meta, helper] = useField(name);
  const { i18n } = useTranslation();
  const lang: Locale = i18n.language as Locale;
  const [stateValue, setStateValue] = useState<string>(formatNumber(field.value, lang));

  // konvertiert zwischen number (für Store) und string (für Komponente)
  const format = (value: number): string =>
    typeof value === 'string' ? value : isNaN(value) ? stateValue : formatNumber(value, lang);
  const parse = (value: string): number => parseNumber(value, lang);
  const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setStateValue(event.target.value);
    helper.setValue(parse(event.target.value));
  };

  return (
    <TextField
      id={id}
      value={format(field.value)}
      onChange={onChange}
      label={label}
      error={!!meta.error}
      helperText={meta.error}
      sx={{ margin: 1 }}
    />
  );
};
