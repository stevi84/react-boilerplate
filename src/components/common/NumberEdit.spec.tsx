import { describe, it, expect, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import { NumberEdit } from './NumberEdit';

const { setValueMock } = vi.hoisted(() => ({ setValueMock: vi.fn() }));
vi.mock('formik', () => ({ useField: () => [{ value: 1.01 }, { error: 'error' }, { setValue: setValueMock }] }));

vi.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (str: string) => str,
    i18n: {
      language: 'de',
      changeLanguage: vi.fn(),
    },
  }),
}));

describe('NumberEdit', () => {
  it('should equal saved snapshot', () => {
    const tree = render(<NumberEdit id="id" name="name" label="label" />).asFragment();
    expect(tree).toMatchSnapshot();
  });

  it('should format the given value', () => {
    render(<NumberEdit id="id" name="name" label="label" />);
    expect(screen.getByLabelText('label')).toHaveValue('1,01');
  });

  it('should parse the input', () => {
    render(<NumberEdit id="id" name="name" label="label" />);
    expect(setValueMock.mock.calls.length).toEqual(0);
    fireEvent.change(screen.getByLabelText('label'), { target: { value: '1,02' } });
    expect(setValueMock.mock.calls.length).toEqual(1);
    expect(setValueMock.mock.calls[0][0]).toEqual(1.02);
  });
});
