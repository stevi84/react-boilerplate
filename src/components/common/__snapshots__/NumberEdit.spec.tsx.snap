// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NumberEdit > should equal saved snapshot 1`] = `
<ForwardRef(TextField)
  error={true}
  helperText="error"
  id="id"
  label="label"
  onChange={[Function]}
  sx={
    {
      "margin": 1,
    }
  }
  value="1,01"
/>
`;
